### DISCUSSION TITLE
Error when assigning a Date on a nullable property

### DISCUSSION THREAD

**Discussion by kalu06** [mikro-orm/mikro-orm#4833] [General]
**Describe the bug**
I have some entity properties declared as follows:
```
@Property({ columnType: 'timestamp(3)', nullable: true })
createdAt: Date | null = null
```
I need this to be able to have `null` values from factories and be able to compare those with entities fetched from the database.

I get an error when trying to assign a date with `orm.em.assign(foo, { createdAt: new Date() });`, although `foo.createdAt = new Date();` works.

(see below to reproduce with a full example)

**Stack trace**
```
ValidationError: Trying to set Foo.createdAt of type 'string' to 'Mon Oct 16 2023 13:58:46 GMT+0200 (Central European Summer Time)' of type 'date'
    at ValidationError.fromWrongPropertyType (/Users/lucascrivelli/Documents/Atipik/mikro-orm-test/node_modules/@mikro-orm/core/errors.js:22:16)
    at EntityValidator.validateProperty (/Users/lucascrivelli/Documents/Atipik/mikro-orm-test/node_modules/@mikro-orm/core/entity/EntityValidator.js:65:44)
    at EntityAssigner.assignProperty (/Users/lucascrivelli/Documents/Atipik/mikro-orm-test/node_modules/@mikro-orm/core/entity/EntityAssigner.js:77:49)
    at /Users/lucascrivelli/Documents/Atipik/mikro-orm-test/node_modules/@mikro-orm/core/entity/EntityAssigner.js:30:35
    at Array.forEach (<anonymous>)
    at EntityAssigner.assign (/Users/lucascrivelli/Documents/Atipik/mikro-orm-test/node_modules/@mikro-orm/core/entity/EntityAssigner.js:29:27)
    at SqlEntityManager.assign (/Users/lucascrivelli/Documents/Atipik/mikro-orm-test/node_modules/@mikro-orm/core/EntityManager.js:989:40)
    at /Users/lucascrivelli/Documents/Atipik/mikro-orm-test/dist/test.js:12:12
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  entity: undefined
}
```
It seems that the type inferred from the EntityValidator is wrong (string instead of date).

**To Reproduce**
1. Create an entity with a property such as described above
```
import { Entity, PrimaryKey, Property } from "@mikro-orm/core";

@Entity()
export class Foo {
  @PrimaryKey()
  id!: number

  @Property()
  bar!: string

  @Property({ columnType: 'timestamp(3)', nullable: true })
  createdAt: Date | null = null
}
```
2. Try to `orm.em.assign` with a date like so:
```
const orm = await MikroORM.init(mikroOrmConfig);
const foo = orm.em.create(Foo, { bar: 'baz' });
orm.em.assign(foo, { createdAt: new Date() }); // doesn't work
// foo.createdAt = new Date(); // works
await orm.em.flush();
await orm.close();
```

**Expected behavior**
`orm.em.assign` to work the same as setting the property value directly.

**Versions**

| Dependency | Version |
| - | - |
| node | 18.6.0 |
| typescript | 5.2.2 |
| mikro-orm | 5.8.9 |
| mikro-mysql | 5.8.9 |

**Comment by B4nan** [comment]
The problem here is that you don't set the `type`, and since you use a union type, that cannot be inferred via reflect-metadata, so the ORM thinks the property is actually a string.

You can do this instead:

```ts
@Property({ type: 'date', columnType: 'timestamp(3)', nullable: true })
createdAt: Date | null = null
```

**Reply by kalu06** [reply]
Great! Works like a charm 🙂.
Thanks again @B4nan for your reactivity

**Reply by B4nan** [reply]
Let me know if it helps, I can try to make this work automatically, as we should be able to infer the type from the `columnType` too (it already works like that for few types, namely the JSON columns).

**Reply by kalu06** [reply]
Would be great to infer this one automatically as well as it's kind of tricky to know we need to specify `type` for this use case, but I'm not sure it's that important. What do you think?

**Reply by B4nan** [reply]
Yeah it's indeed best if things just work :] I was looking at the code and we already do this for bigints and JSON props, it's just about adding 3 lines somewhere in this method:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/metadata/MetadataDiscovery.ts#L992

**Reply by tobiaskraus** [reply]
I had the same issue, and the `type: "date"` solution worked for me. But wanted to mention, that it only crashed while updating from `@5.4.2` to `@5.8.8`. Somehow it wasn't an issue in older mikro-orm version.

**Reply by B4nan** [reply]
This will do it https://github.com/mikro-orm/mikro-orm/commit/6bc116aabdc9ca958238faaa7ac4a44cf3c71c08?w=1

**Reply by B4nan** [reply]
> and the type: "date" solution worked for me

FYI in v6 `type: 'date'` means date, not datetime, so if you want to be future proof, use `type: 'datetime'` or `type: 'Date'` (as `Date` is how it would be normally inferred.

(https://github.com/mikro-orm/mikro-orm/blob/v6/docs/docs/upgrading-v5-to-v6.md#changes-in-date-property-mapping)
